### PR TITLE
Notify user if they are using an outdated version of Manim

### LINKED ISSUE #1230: Inform Manim Community Users when using an out of date version
## Enhancement proposal

This came up in a discussion on [discord](https://discord.com/channels/581738731934056449/581738732646957057/827411746318843924), it would be nice to let users know when a new released version is available -- or just that they're no longer up to date with the most recent tagged/stable version -- just like pip does.

I'm not sure how this would be implemented for versions prior to v0.5.0, but ideally this warning/message would be made on each execution of the `manim` commmand. Additionally, the user should be able to silence this warning/label via the config files if they really don't want to update for whatever reason.

### PR DESCRIPTION
**Pull request by RickyC0626**
## Changelog / Overview

Closes #1230 

## Motivation

Notifies users in a non-obtrusive way if their local version is outdated and if there is an update available.

## Explanation for Changes

After rendering, `manim` will notify the user if their installed version is outdated. This feature is also available as a global option under `manim render -h` named `--notify_outdated_version`.

When the installed version is outdated:
image

## Testing Status

System: WSL Ubuntu-20.04

## Further Comments

We can test this non-obtrusive method to see if it improves user experience. If it doesn't, then we can aim for a more aggressive approach.

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

File: manim/__main__.py
Comment (naveen521kk) [inline manim/__main__.py:3]: Is requests dependency added?

File: manim/__main__.py
Comment (naveen521kk) [inline manim/__main__.py:20]: I would suggest using PyPi API https://pypi.org/pypi/manim/json as it isn't ratelimited.

File: manim/__main__.py
Comment (naveen521kk) [inline manim/__main__.py:28]: I would like to be printed or the request being made only at the end of the animation and it should be disabled using config files or cli options.

File: manim/__main__.py
Comment (jsonvillanueva) [reply manim/__main__.py:3]: This would need to be added to `pyproject.toml`

File: manim/__main__.py
Comment (naveen521kk) [reply manim/__main__.py:20]: Also, I would be able to run Manim offline, the error need to be handled.

File: manim/__main__.py
Comment (friedkeenan) [reply manim/__main__.py:3]: Fairly sure manim already has a dependency on requests through one of our other dependencies

File: manim/__main__.py
Comment (RickyC0626) [reply manim/__main__.py]: Would you like this feature to be enabled or disabled by default?

Comment (jsonvillanueva) [reply manim/__main__.py]: Definitely enabled by default -- perhaps the warning message should contain a short blurb about where to disable this settings in the future, but I'm okay with not having a short blurb for that.

File: manim/__main__.py
Comment (naveen521kk) [reply manim/__main__.py:3]: > Fairly sure manim already has a dependency on requests through one of our other dependencies

anyhow we need to add it though.

File: manim/__main__.py
Comment (kolibril13) [reply manim/__main__.py]: For people with a slow internet connection:
Would this feature make everytime a web request when the script is running?
And if yes: Would that not result in a longer run time of a script?

Comment (naveen521kk) [reply manim/__main__.py]: That's why it should be run at the end of the animation, like how pip is doing. They can disable if they want to.

Comment (kolibril13) [reply manim/__main__.py]: Sounds good. what about the parameter `check_for_new_version = True` at this point:
https://github.com/ManimCommunity/manim/blob/2752ca54d7437371aeecad57302d75189cf7b0de/manim/_config/default.cfg#L16
?

Comment (RickyC0626) [reply manim/__main__.py]: Yup, that's what I'm planning to do. I've named it `notify_outdated_version`.

File: manim/cli/render/commands.py
Comment (naveen521kk) [inline manim/cli/render/commands.py:154]: just say config file, don't say `manim/_config/default.cfg`, it can be in the same directory or user directory also.

Comment (RickyC0626) [reply manim/cli/render/commands.py:154]: Oh right, just realized that.

File: manim/cli/render/global_options.py
Comment (jsonvillanueva) [inline manim/cli/render/global_options.py:37]:
```suggestion
        help="Display warnings for outdated installation.",
```

File: pyproject.toml
Comment (naveen521kk) [inline pyproject.toml:52]: update lock file? You should run `poetry lock` command after adding this.

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py]: No need to inform users they're up to date. Especially because this isn't always true... they would be running the latest released version (e.g. v0.5.0 for now), but the latest is actually whatever is on master
```suggestion
```

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py]: I think @naveen521kk already mentioned that this should be a general config file. Users shouldn't modify default.cfg since that's overwritten on update -- they should use their own folder-wide/custom config file

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:25]: I don't think this should be placed here

**Review by jsonvillanueva — CHANGES_REQUESTED** [review]

**Comment by jsonvillanueva** [comment]
Please don't force push once there have been reviews on your PR -- I'll make a PR noting this in our contributing guidelines -- but this makes it difficult to see what's already been reviewed since we last reviewed (especially if the PR is very large -- this one isn't so it's not that big a deal).

**Comment by RickyC0626** [comment]
> Please don't force push once there have been reviews on your PR -- I'll make a PR noting this in our contributing guidelines -- but this makes it difficult to see what's already been reviewed since we last reviewed (especially if the PR is very large -- this one isn't so it's not that big a deal).

Ah alright

File: docs/source/contributing/reference/manim_directive.rst
Comment (naveen521kk) [inline docs/source/contributing/reference/manim_directive.rst:1]: unrelated change?

Comment (RickyC0626) [reply docs/source/contributing/reference/manim_directive.rst:1]: I believe that was autogenerated for some odd reason, when opening rst files.

File: manim/__main__.py
Comment (naveen521kk) [inline manim/__main__.py:3]: unecessary import

File: manim/cli/render/commands.py
Comment (WampyCakes) [inline manim/cli/render/commands.py:153]: I think that this is a good feature until right here. This print (along with the one below it with the link) are longer and more obtrusive than just always displaying if manim is outdated. If it were me receiving the warning, I'd probably keep it around just so that I don't forget I'm not on the most recent version, but because of the long message, I'd probably turn it off.

Comment (WampyCakes) [reply manim/cli/render/commands.py:153]: I'm pretty sure I clicked the right line, but oh well, _github_. Since it isn't clear what I am talking about, I am referencing notifying the user how to turn off the notification (this seems like a weird UX choice anyway).

Comment (jsonvillanueva) [reply manim/cli/render/commands.py:153]: I guess the installed/latest version could be combined into a single line similar to how pip does it. Taking from stackoverflow:
> You are using pip version 6.0.8, however version 8.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

I think the above example with pip should be translated to manim as:
>You are using manim version A.B.C, however version D.E.F is available.
You should consider upgrading via `pip install -U manim`.

Additionally, pip also has a config file setting to disable/quiet this warning, but they don't mention it. I think it's fine to keep the warning but I'm not sure about the link to the latest documentation. Again, `latest != most recently released`

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:158]:
```suggestion
```

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:152]: Combine into one line as per the example with pip

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:108]: I'm pretty sure you can just reference the config dict for the if condition after rendering on L42
```suggestion
```

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:142]: like so:
```suggestion
    if config.notify_outdated_version:
```

File: manim/cli/render/commands.py
Comment (naveen521kk) [reply manim/cli/render/commands.py:158]: use `stable`

Comment (jsonvillanueva) [reply manim/cli/render/commands.py:158]: I still don't think we should link to this website in the trace. Once `manim cfg` is made better, just like pip has the option to disable the warning with:
`pip config set global.disable-pip-version-check true`
it'd be better to just tell them what `manim cfg` command should be run to disable the warning.

File: manim/cli/render/commands.py
Comment (WampyCakes) [reply manim/cli/render/commands.py:153]: I just think it's counterproductive to notify a user on how to turn off another notification. It takes far more space in the terminal than just mentioning you're outdated. And if it's combined into one line, I especially don't see a reason for it. We could remove this message and just mention it in our docs so that if someone searches how to disable the warning (which I already doubt will happen if it's a small warning), they'll find mention of it in our docs.

File: manim/cli/render/commands.py
Comment (RickyC0626) [reply manim/cli/render/commands.py:108]: > I'm pretty sure you can just reference the config dict for the if condition after rendering on L142

`config.digest_args(click_args)` on L109 considers no arguments entered as `False` for `notify_outdated_version`

Comment (RickyC0626) [reply manim/cli/render/commands.py:108]: Perhaps we should remove it as a flag and keep it as `True` by default? Then later on we can add it back if necessary.

Comment (jsonvillanueva) [reply manim/cli/render/commands.py:108]: In practice, passing the CLI flag should overwrite the config files and passing no CLI flag shouldn't modify the ManimConfig object. Since it does, that's a bug, though I'm unsure if it's click/cloup related or ManimConfig related -- I'll checkout your branch and see what I can do to move this PR along; but I'd prefer to keep the flag around.

Comment (RickyC0626) [reply manim/cli/render/commands.py:108]: I believe it has something to do with the changes to `manim/_config/utils.py`

Comment (jsonvillanueva) [reply manim/cli/render/commands.py:108]: This works in global_options.py:
```
    option(
        "--notify_outdated_version/--silent",
        is_flag=True,
        default=None,
        help="Display warnings for outdated installation.",
    ),
```
So the flag takes 3 values, False/True if specified, otherwise None so that the config file can do its job

**Review by jsonvillanueva — APPROVED** [review]
LGTM

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:140]: still, `args["notify_outdated_version"]` isn't actually necessary here if the config is doing it's job properly

Comment (RickyC0626) [reply manim/cli/render/commands.py:140]: The user may want the default to be `False` and explicitly check using the flag.

Comment (jsonvillanueva) [reply manim/cli/render/commands.py:140]: The config file will handle the explicit check of the flag if it's passed in.
Suppose `manim.cfg` (users shouldn't modify `default.cfg` as it will change on update) has:
```
[CLI]
notify_outdated_version = False
```
Running `manim -pql ...` will render the file and mute the warning from being outdated due to the presence of a manim.cfg file.
If for some reason the user needed to run `manim -pql --notify_outdated_version ... ` (wow that's a long flag name btw), the `config.digest_arg(args)` before any of the rendering will handle the case of explicitly checking the flag. Again, if it doesn't it's a bug.

Comment (jsonvillanueva) [reply manim/cli/render/commands.py:140]: Also, as you might have noticed, some of the other flags in `@global_options, @output_options, @render_options, @ease_of_access_options` are missing the `default = None` that was critical to this option being usable with the config.
This is a relatively easy bug fix that should be done in a separate PR, but really the CLI just needs more exhaustive test cases.

File: manim/cli/render/commands.py
Comment (behackl) [inline manim/cli/render/commands.py:142]: Two comments regarding these two lines:

- I'm not particularly fond of a variable named `manim` and would prefer if it were renamend (I won't block the PR over it though),
- Given that `requests.get` raises an exception when the request fails (`pypi.org` could be down, or no internet connection could be available), I think it should be handled and a warning should be logged instead informing the user that it could not be checked whether the installed version is the latest. (I think this is actually important and I'd like to have that implemented before merging.)

**Review by behackl — DISMISSED** [review]
Thank you for your contribution! I have a minor change request that I'd like to have discussed, but otherwise this seems good to go.

File: manim/cli/render/commands.py
Comment (jsonvillanueva) [inline manim/cli/render/commands.py:160]: This one has a `.` instead of a `,`
```suggestion
            logger.warning("Something else went wrong: %s",warn_prompt)
```
But generally I would prefer using f-strings here (non-blocking):
```suggestion
            logger.warning(f"Something else went wrong: {warn_prompt}")
```

Comment (behackl) [reply manim/cli/render/commands.py]: This doesn't seem to be resolved yet.

Comment (RickyC0626) [reply manim/cli/render/commands.py]: Just pushed 👍🏻

Comment (jsonvillanueva) [reply manim/cli/render/commands.py]: > 
> 
> This doesn't seem to be resolved yet.

I only cared about the actual error with `,` vs `.`, but f-strings are appreciated!

File: manim/__main__.py
Comment (WampyCakes) [inline manim/__main__.py:18]: Why wasn't the newline necessary before?

File: manim/cli/render/commands.py
Comment (WampyCakes) [inline manim/cli/render/commands.py:157]:
```suggestion
            logger.warning(f"Timed Out: {warn_prompt}")
```

File: manim/cli/render/commands.py
Comment (WampyCakes) [inline manim/cli/render/commands.py:160]:
```suggestion
            logger.warning(f"Something went wrong: {warn_prompt}")
```

File: manim/cli/render/commands.py
Comment (WampyCakes) [inline manim/cli/render/commands.py:144]: Since this is only used in one place, you could just remove the `releases` variable and plug it straight into `max()` to get the stable release.

File: manim/__main__.py
Comment (RickyC0626) [reply manim/__main__.py:18]: It was more of a personal preference, having a space between the title and the descriptions.

File: manim/cli/render/commands.py
Comment (naveen521kk) [inline manim/cli/render/commands.py:160]: Catching error is fine: you shouldn't raise or exit because Manim should run without internet also.

File: manim/cli/render/commands.py
Comment (naveen521kk) [inline manim/cli/render/commands.py:147]: don't set a timeout manually 
```suggestion
            req_info = requests.get(manim_info_url)
```

File: manim/cli/render/commands.py
Comment (naveen521kk) [inline manim/cli/render/commands.py:165]: looking at https://pypi.org/pypi/manim/json it seems it is better to directly pick version from info rather than doing this

```suggestion
        try:
            releases = req_info.json()["info"]["version"]

            if releases != __version__:
```

File: manim/cli/render/commands.py
Comment (naveen521kk) [inline manim/cli/render/commands.py:175]: Again, doing exit or raise, this feature shouldn't make Manim exit with an error code.

Comment (naveen521kk) [reply manim/cli/render/commands.py:175]: Also, it should use `logger.warning` rather it should use `logger.debug` because the user doesn't need to know that this failed.

File: manim/__main__.py
Comment (WampyCakes) [reply manim/__main__.py:18]: I'm okay with that

**Review by WampyCakes — APPROVED** [review]
LGTM! Good add!

**Review by naveen521kk — APPROVED** [review]
lgtm

**Pull request merged**
